subprojects/model-core/src/main/java/org/gradle/model/internal/core/NodeBackedModelSet.java

```diff
@@ -165,9 +170,39 @@ public void clear() {
         throw new UnsupportedOperationException();
     }
 
-    // TODO - mix this in using decoration. Also validate closure parameter types, if declared
+    // TODO - mix this in using decoration.
     public void create(Closure<?> closure) {
-        create(ClosureBackedAction.of(closure));
+        create(getElementTypeClass(), closure);
+    }
+
+    public void create(Class<? extends T> type, Closure<?> closure) {
+        if (hasNoParameters(closure)) {
+            create(ClosureBackedAction.of(closure));
+        } else if (hasOneParameter(closure)) {
+            validateClosureParameter(type, closure);
+            create(type, ClosureBackedAction.of(closure));
+        } else {
+            throw new TooManyClosureParametersException(type);
```

Comment:
This should throw an `IllegalArgumentException` instead to be on par with what `ModelMap` does, see [NodeBackedModelMap.java#L528](https://github.com/gradle/gradle/blob/f490bdf61bd9b4f5383cd9fb0d8ffbca93da8c32/subprojects/model-core/src/main/java/org/gradle/model/internal/core/NodeBackedModelMap.java#L528)